File: Core/Core/AppEnvironment/ExperimentalFeature.swift
//
// This file is part of Canvas.
// Copyright (C) 2019-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

import Foundation

/// An experimental or in-development feature
///
/// An `ExperimentalFeature` flag is useful for including code in production
/// that is not ready for all users to exercise. This is different from
/// feature flags in Canvas, which represent optional functionality in
/// production that should only apply to certain accounts, courses, or people.
public enum ExperimentalFeature: String, CaseIterable, Codable {
    case favoriteGroups = "favorite_groups"
    case nativeStudentInbox = "native_student_inbox"
    case nativeTeacherInbox = "native_teacher_inbox"
    case K5Dashboard = "enable_K5_dashboard"
    case hybridDiscussionDetails = "hybrid_discussion_details"

    public var isEnabled: Bool {
        get {
            // If there are no saved values for K5 mode we return true by default. Debug builds without Firebase feature flag fetch will have this enabled.
            if self == .K5Dashboard, UserDefaults.standard.object(forKey: userDefaultsKey) == nil {
                return true
            }
            return UserDefaults.standard.bool(forKey: userDefaultsKey)
        }
        nonmutating set { UserDefaults.standard.set(newValue, forKey: userDefaultsKey) }
    }

    public var userDefaultsKey: String {
        return "ExperimentalFeature.\(self.rawValue)"
    }

    public static var allEnabled: Bool {
        get { ExperimentalFeature.allCases.allSatisfy({ $0.isEnabled }) }
        set { ExperimentalFeature.allCases.forEach({ $0.isEnabled = newValue }) }
    }
}

File: rn/Teacher/src/common/ExperimentalFeature.js
//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */
import { Settings } from 'react-native'

export default class ExperimentalFeature {
  constructor (remoteConfigKey) {
    this.remoteConfigKey = remoteConfigKey
    this.settingsKey = `ExperimentalFeature.${remoteConfigKey}`
    this.enabled = Boolean(Settings.get(this.settingsKey))
  }

  get isEnabled () {
    return this.enabled
  }

  set isEnabled (value) {
    this.enabled = value
    Settings.set({ [this.settingsKey]: value })
  }

  static get allEnabled () {
    return Object.values(ExperimentalFeature).every(f => f.isEnabled)
  }

  static set allEnabled (newValue) {
    Object.values(ExperimentalFeature)
      .forEach(f => { f.isEnabled = newValue })
  }
}

// *** Please stay in sync with Core/AppEnvironment/ExperimentalFlags.swift ***
// There is no automatic syncing of individual flags, since they can't be async
// and should be static.
ExperimentalFeature.favoriteGroups = new ExperimentalFeature('favorite_groups')
ExperimentalFeature.nativeStudentInbox = new ExperimentalFeature('native_student_inbox')
